### DISCUSSION TITLE
Scripts don't have the right CORS headers when deployed to Vercel

### DISCUSSION THREAD

**Discussion by m11m** [vercel/next.js#41207] [Help] [answered]
Hey guys, I have the following headers set in my next config. I'm using a `SharedArrayBuffer` on a few pages, and in order to use `SharedArrayBuffer`, browsers require these cross origin settings to be used.

This works correctly in on my local machine (in both dev and prod mode). It works when these headers are set, and throws an error if I remove them from the config.

~~The problem is that when I deploy my site to Vercel, these headers aren't used on anything, and the site throws an error.~~

~~Not sure why they aren't being used, would appreciate any input~~

```js
  async headers() {
    return [
      {
        source: "/:path*",
        headers: [
          {
            key: "Cross-Origin-Opener-Policy",
            value: "same-origin",
          },
          {
            key: "Cross-Origin-Embedder-Policy",
            value: "require-corp",
          },
        ],
      },
    ];
  },
```

UPDATE:

It looks like I had a bit of a misunderstanding.

The headers ARE being used, just not on the worker scripts, so they throw an `ERR_BLOCKED_BY_RESPONSE` error when I try to import them.

![image](https://user-images.githubusercontent.com/92771507/194261631-0374e01c-09e9-4a5f-ad3e-fe7a3b26a955.png)

![image](https://user-images.githubusercontent.com/92771507/194401213-e506ce89-a44b-40cf-8320-e66cd4de3d81.png)

The error says iframe, but im trying to load a worker. I read this as I need to add a `Cross-Origin-Embedder-Policy` header to the scripts im trying to import. On my local machine they have this header and it works, but for some reason it doesn't on the deployed site.

**Comment by icyJoseph** [comment]
Odd, I am not sure why the headers would not be honoured, but you could create a `vercel.json` at the root of your project and put those headers there, see if that works. Sort of like it is done here: https://blakey.co/blog/setting-up-security-headers-with-vercel 

## EDIT

The headers were working fine, OP was able to narrow down the issue to something else.

**Reply by m11m** [reply]
using `vercel.json` still had the same problem :/
I updated my post with some more information, I was able to narrow it down a bit. Not sure how helpful it is.

**Reply by icyJoseph** [reply]
Isn't this a kind of situation where the resource you are loading has to give you the permission? I guess you are in control of the worker script?

**Reply by m11m** [reply]
That's how I understand it. I'm in control of the worker script, I create it like this:

```js
new Worker(new URL("./worker.ts", import.meta.url), { type: "module" });
```

**Reply by icyJoseph** [reply]
This is but a cop-out but could you implement middleware, just to test a possible solution?

```ts
import { NextResponse, type NextRequest } from "next/server";

export function middleware(req: NextRequest) {
  const { pathname } = req.nextUrl;

  if (pathname.startsWith("/_next/static/chunks/worker")) {
    const response = NextResponse.next();

    response.headers.set("Cross-Origin-Embedder-Policy", "require-corp");

    return response;
  }

  return NextResponse.next();
}

export const config = {
  matcher: ["/_next/static/chunks/:rest*"],
};

```

And then make your worker something like:

```tsx
useEffect(() => {
    workerRef.current = new Worker(
      new URL(/* webpackChunkName: 'worker' */ "../worker.ts", import.meta.url)
    );
    workerRef.current.onmessage = (event: MessageEvent<number>) =>
      alert(`WebWorker Response => ${event.data}`);

    const currentWorker = workerRef.current;

    return () => {
      currentWorker.terminate();
    };
  }, []);

  const handleWork = useCallback(async () => {
    workerRef.current?.postMessage(100000);
  }, []);
```

The important bit being the `webpackChunkName` magic comment.

How does this do? Perhaps other headers are needed as well.

**Reply by m11m** [reply]
Doesn't seem to work, still get the same result. I tested it locally and confirmed the middleware is applying the header, and the scripts have the header when I import them, but on the deployment no luck.

**Reply by icyJoseph** [reply]
Do you mind sharing a link to the deployment? This is very strange. If `vercel.json` also doesn't quite make it, then I'd guess you should contact the Vercel support team to try and get a bit more insight. So far, it doesn't seem to be an issue with the framework, but rather with how the Vercel CDN is handling the scripts.

**Reply by m11m** [reply]
yeah you should be able to view it if you go here https://client-rk667w99a-wired.vercel.app/app/test

I created a help ticket so hopefully they can shed some light as to whats happening here. Its not specific to the worker scripts, none of the scripts have the headers I set. Its just that the worker scripts require the headers to be imported so they throw an error.

Thanks for helping me out btw I really appreciate it 🙏

**Comment by m11m** [comment, marked as answer]
I think I found the solution / problem! If you remove the `i18n` section from the next config it seems to work fine.

I created an example repository for it:
https://github.com/0xlain/vercel-cors-bug

**Reply by icyJoseph** [reply]
Oh wow, good job! but that only makes the mystery more mysterious! 😅 Good stuff

**Reply by m11m** [reply]
@Inhance-Dev no
